cpp/ql/src/Likely Bugs/Underspecified Functions/MistypedFunctionArguments.ql

```diff
@@ -0,0 +1,109 @@
+/**
+ * @name Call to a function with one or more incompatible arguments
+ * @description A call to a function with at least one argument whose type does
+ * not match the type of the corresponding function parameter.  This may indicate
+ * that the author is not familiar with the function being called.  Passing mistyped
+ * arguments on a stack may lead to unpredictable function behavior.
+ * @kind problem
+ * @problem.severity warning
+ * @precision very-high
+ * @id cpp/mistyped-function-arguments
+ * @tags correctness
+ *       maintainability
+ */
+
+import cpp
+
+pragma[inline]
+int sizeofInt() { result = any(IntType pt).getSize() }
+
+pragma[inline]
+predicate pointerArgTypeMayBeUsed(Type arg, Type parm) {
+  arg = parm
+  or
+  // arithmetic types
+  arg instanceof ArithmeticType and
+  parm instanceof ArithmeticType and
+  arg.getSize() = parm.getSize() and
+  (
+    (
+      arg instanceof IntegralType and
+      parm instanceof IntegralType
+    )
+    or
+    (
+      arg instanceof FloatingPointType and
+      parm instanceof FloatingPointType
+    )
+  )
+  or
+  // pointers to void are ok
+  arg instanceof VoidType
+  or
+  parm instanceof VoidType
+}
+
+pragma[inline]
+predicate argTypeMayBeUsed(Type arg, Type parm) {
```

Comment:
This predicate doesn't fully model [the definition of _compatible types_](https://en.cppreference.com/w/c/language/type#Compatible_types), and omissions will lead to false positives. In particular, the predicate is missing support for enums, structs, and functions. I suggest that you support structs and functions by just comparing structs by size for now and assuming all function pointers are interchangeable. We can always refine it, but lack of support should lead to false negatives rather than false positives.

Reply:
I would argue for leaving things as they are, i.e., being strict about <code>enum</code> and <code>struct></code> compatibility (as captured by the <code>arg = parm</code> clause of <code>argTypeMayBeUsed</code>).  We did not hit any (false) positives in the 68-project query.  (If you can think of an even bigger set, please share it with me.)